Refs #35730: Signed 'uid’ parameter instead of base64-encoding to prevent possible user count leakage

lapinvert opened the pull request:
#### Trac ticket number

[ticket-35730](https://code.djangoproject.com/ticket/35730)

#### Branch description
When using Django’s default view for requesting a password reset, `PasswordResetView`, the `PasswordResetForm`’s `save()` method sends an email containing a `uid` parameter generated using `urlsafe_base64_encode(force_bytes(user.pk))`.

This results in the user’s email inbox containing a password reset link that indirectly reveals the user’s primary key (`user.pk`), which exposes information about how many users exist on any Django site that uses this default view.

Surely, organizations that design their entities with non-enumerable public identifiers (such as by using a `UUIDField` for the primary key) would not be affected by this, however as the issue is also addressed by other means, such as a secondary public identifier, or simply a careful app design, I would still think that many Django site owners who prefer to keep this information private are likely unaware that it’s being exposed through this native mechanism.

To prevent the leakage of the user.pk value by default, I replaced the base64 encoding with the signing of the `user.pk` value.

#### Checklist
- [x] This PR targets the `main` branch.
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

nessita commented:
Hello @lapinvert, thank you for the report and PR solving it! This is a great start :star2: 

I haven't done a in-depth review yet but there are a few things that you could progress before that:
1. Fixing the lint errors (note that while code line length is 88 cols, comments line length is 79cols). More details in [these code style docs](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#coding-style).
2. This branch should provide a small adding in the release notes describing the change.

Also, have you thought about what will happen to those password reset links that were sent prior this change but haven't been used nor expired yet? I think we need to evaluate providing a compatibility path where both type of links work for a while.

lapinvert commented:
Hi @nessita!
Thank you so much for the feedback.

I’ve pushed updates to the branch addressing your points as well as a related issue from the ticket discussion:
- Transitioned from signing to light encryption since signing would not prevent the leakage
- Fixed the linting issues
- Added a description of the change in the release notes
- Implemented temporary support for the old behavior to allow for a smoother transition

Let me know if there’s anything else you’d like me to adjust or improve!

nessita commented on django/contrib/auth/tokens.py:
```diff
@@ -128,5 +135,40 @@ def _now(self):
         # Used for mocking in tests
         return datetime.now()
 
+    def _xor_encrypt_decrypt(self, uid):
```
The two "public" APIs being added to this method should be fully tested, including the corner cases described in the docstring (thanks for the thorough details).

nessita commented on django/contrib/auth/tokens.py:
```diff
@@ -128,5 +135,40 @@ def _now(self):
+
+    def encrypt_uid(self, uid):
+        """
+        Returns a XOR-encrypted user id for use in the password reset mechanism.
```
Docstrings should follow PEP-8 and PEP-257, specifically verbs should be written as a command. An alternative proposal would be:
```suggestion
        XOR-encrypt the provided user ID and return its base64-encoded representation.
```

nessita commented on django/contrib/auth/tokens.py:
```diff
@@ -128,5 +135,40 @@ def _now(self):
+
+    def decrypt_uid(self, encrypted_uidb64):
+        """
+        Returns the decrypted user id given the base64-encoded encrypted user id.
```
```suggestion
        Return the user id given its base64-encoded XOR-encrypted representation.
```

nessita commented on django/contrib/auth/views.py:
```diff
@@ -309,7 +309,22 @@ def get_user(self, uidb64):
-            user = None
+            pass
+
+        # Temporarily support the old format of uid base64-encoded as plain text
```
This is good, supporting old links. But we can't have this code forever in Django, we need a clear and documented deprecation path. My advice is to follow the [deprecating a feature docs](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature) and raise a deprecation warning for each successful hit for old-style links.

I do wonder if we need a more clever approach here to handle the deprecation... There is an [ongoing conversation](https://forum.djangoproject.com/t/signing-the-csrf-cookie/35156/5) about signing CSRF cookies which has some common points with this. I will reference this PR in the forum topic but at this point I think we need to involve the community to design a good plan for this.

nessita replied:
Also, whatever we end up doing for the deprecation/transition of link and token obfuscation, we need to add tests for it.

nessita commented on docs/releases/5.2.txt:
```diff
@@ -52,6 +52,8 @@ Minor features
 * The default iteration count for the PBKDF2 password hasher is increased from
   870,000 to 1,000,000.
 
+* The ``uid`` parameter in password reset emails is now obfuscated to prevent potential leakage of user count. This change hides the user's primary key (``user.pk``), which was previously only base64-encoded. The obfuscation is achieved using a simple XOR cipher.
```
All the docs and docstrings lines should be wrapped at 79cols. Could you please summarize this a bit and ensure it wraps at that line length?

nessita reviewed (changes requested):
Hello @lapinvert, thank you for this work! I value the effort and detailed docstrings.

I have added a fair amount of comments, mostly focused one:
1. Some tests are missing.
2. We need to improve the deprecation approach for old links (unsure how).

As mentioned in one of the comment, at this point I think we need to involve more people from the community, so writing a forum post would be ideal. What do you think?
